pkg/controllers/management/cloudcredential/controller.go

```diff
@@ -2,32 +2,163 @@ package cloudcredential
+	harvesterCloudCredentialExpirationAnnotation    = "rancher.io/expiration-timestamp"
+)
+
 type Controller struct {
 	managementContext *config.ManagementContext
+	secretClient      corecontrollers.SecretClient
+	tokenClient       mgmtcontrollers.TokenClient
+	apply             apply.Apply
 }
 
-func Register(ctx context.Context, management *config.ManagementContext) {
+func Register(ctx context.Context, management *config.ManagementContext, clients *wrangler.Context) {
 	m := Controller{
 		managementContext: management,
+		secretClient:      clients.Core.Secret(),
+		tokenClient:       clients.Mgmt.Token(),
+		apply: clients.Apply.WithCacheTypes(
+			clients.Core.Secret(),
+			clients.Mgmt.Token(),
+		),
 	}
 	management.Core.Secrets("").AddHandler(ctx, "management-cloudcredential-controller", m.ccSync)
+	if features.Harvester.Enabled() {
+		clients.Core.Secret().OnChange(ctx, "harvester-cloud-credential-token", m.syncHarvesterToken)
+	}
+}
+
+// syncHarvesterToken will extend the ttl of a token owned by a harvester cloud credential to infinite, as well as
+// handle deletion of tokens when the cloud credential secret is deleted. Under normal circumstances, kubeconfigs
+// tokens are subject to the settings.KubeconfigDefaultTokenTTLMinutes setting, which denotes a max ttl of default 30
+// days. This does not make for a particularly satisfying UX for administrators of Harvester provisioned node driver
+// clusters, as cloud credentials require manual rotation or be subject to failures during machine creation/deletion.
+// Deletion is managed by this handler via wrangler's apply mechanism, which is used to indicate the secret is the
+// owner of the token, sans owner references. A finalizer is applied to the secret, which will be used to delete the
+// now infinitely lived tokens when the cloud credential itself is deleted, preventing a resource leak within the
+// local cluster.
+func (c *Controller) syncHarvesterToken(key string, secret *v1.Secret) (*v1.Secret, error) {
+	if secret == nil || secret.Namespace != namespace.GlobalNamespace {
+		return secret, nil
+	}
+
+	if len(secret.Data["harvestercredentialConfig-kubeconfigContent"]) == 0 {
+		// not a (valid) harvester cloud credential
+		return secret, nil
+	}
+
+	var err error
+
+	d := sha256.Sum256(secret.Data["harvestercredentialConfig-kubeconfigContent"])
+	checksum := hex.EncodeToString(d[:])
+
+	if secret.DeletionTimestamp == nil && secret.Annotations[harvesterCloudCredentialTokenChecksumAnnotation] == checksum {
+		// token TTL has already been extended and tokens have not been rotated
+		return secret, nil
+	} else if secret.DeletionTimestamp != nil && !slices.Contains(secret.Finalizers, harvesterCloudCredentialFinalizer) {
+		// token TTL was never extended
+		return secret, nil
+	}
+
+	objs := make([]runtime.Object, 0)
+
+	if secret.DeletionTimestamp != nil {
+		err = c.apply.WithOwner(secret).ApplyObjects(objs...)
```

Comment:
This is probably a wrangler-ism that I'm not aware of, but what does applying an empty slice of objects do here?

Reply:
Since we create tokens with an `ownerReference` pointing to the secret, I believe this change helps clean those up automatically, which should then allow us to safely remove the finalizer. Jake might have more details to add.